x/ccv/provider/types/msg.go

```diff
@@ -227,11 +234,176 @@ func (msg MsgSubmitConsumerDoubleVoting) GetSigners() []sdk.AccAddress {
+		PowerShapingParameters:   powerShapingParameters,
+	}, nil
+}
+
+// Type implements the sdk.Msg interface.
+func (msg MsgCreateConsumer) Type() string {
+	return TypeMsgCreateConsumer
+}
+
+// Route implements the sdk.Msg interface.
+func (msg MsgCreateConsumer) Route() string { return RouterKey }
+
+// ValidateBasic implements the sdk.Msg interface.
+func (msg MsgCreateConsumer) ValidateBasic() error {
+	if err := ValidateField("chain id", msg.ChainId, cmttypes.MaxChainIDLen); err != nil {
+		return err
+	}
+
+	if err := ValidateConsumerMetadata(msg.Metadata); err != nil {
+		return err
+	}
+
+	if msg.InitializationParameters != nil {
+		if err := ValidateInitializationParameters(*msg.InitializationParameters); err != nil {
+			return err
+		}
+	}
+
+	if msg.PowerShapingParameters != nil {
+		if msg.PowerShapingParameters.Top_N != 0 {
+			return fmt.Errorf("cannot create a Top N chain through `MsgCreateConsumer`; " +
+				"first create the chain and then use `MsgUpdateConsumer` to make the chain Top N")
+		}
+		if err := ValidatePowerShapingParameters(*msg.PowerShapingParameters); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Type implements the sdk.Msg interface.
+func (msg MsgCreateConsumer) GetSignBytes() []byte {
+	bz := ccvtypes.ModuleCdc.MustMarshalJSON(&msg)
+	return sdk.MustSortJSON(bz)
+}
+
 // GetSigners implements the sdk.Msg interface. It returns the address(es) that
 // must sign over msg.GetSignBytes().
-// If the validator address is not same as delegator's, then the validator must
-// sign the msg as well.
-func (msg *MsgConsumerAddition) GetSigners() []sdk.AccAddress {
+func (msg MsgCreateConsumer) GetSigners() []sdk.AccAddress {
+	valAddr, err := sdk.ValAddressFromBech32(msg.Signer)
+	if err != nil {
+		// same behavior as in cosmos-sdk
+		panic(err)
+	}
+	return []sdk.AccAddress{valAddr.Bytes()}
+}
+
+// NewMsgUpdateConsumer creates a new MsgUpdateConsumer instance
+func NewMsgUpdateConsumer(signer string, consumerId string, ownerAddress string, metadata *ConsumerMetadata,
+	initializationParameters *ConsumerInitializationParameters, powerShapingParameters *PowerShapingParameters) (*MsgUpdateConsumer, error) {
+	return &MsgUpdateConsumer{
+		Signer:                   signer,
+		ConsumerId:               consumerId,
+		NewOwnerAddress:          ownerAddress,
+		Metadata:                 metadata,
+		InitializationParameters: initializationParameters,
+		PowerShapingParameters:   powerShapingParameters,
+	}, nil
+}
+
+// Type implements the sdk.Msg interface.
+func (msg MsgUpdateConsumer) Type() string {
+	return TypeMsgUpdateConsumer
+}
+
+// Route implements the sdk.Msg interface.
+func (msg MsgUpdateConsumer) Route() string { return RouterKey }
+
+// ValidateBasic implements the sdk.Msg interface.
+func (msg MsgUpdateConsumer) ValidateBasic() error {
+	if err := ValidateConsumerId(msg.ConsumerId); err != nil {
+		return err
+	}
+
+	const maxNewOwnerAddress = 500
```

Comment:
This shouldn't not be validated like this. See https://github.com/cosmos/cosmos-sdk/blob/8bfcf554275c1efbb42666cc8510d2da139b67fa/x/staking/keeper/msg_server.go#L161 for an example.

Reply:
Thanks!
Your points seems to be for validator addresses. I used:
```
		_, err := sdk.AccAddressFromBech32(msg.NewOwnerAddress)
```
for general accounts. If that's okay, feel free to resolve.

Reply:
Then use this example from MsgSend:
 
- https://github.com/cosmos/cosmos-sdk/blob/8bfcf554275c1efbb42666cc8510d2da139b67fa/proto/cosmos/bank/v1beta1/tx.proto#L46
- https://github.com/cosmos/cosmos-sdk/blob/8bfcf554275c1efbb42666cc8510d2da139b67fa/x/bank/keeper/msg_server.go#L36